crates/sui-graphql-rpc/src/types/validator.rs

```diff
@@ -13,16 +21,73 @@ use super::owner::Owner;
 use super::sui_address::SuiAddress;
 use super::validator_credentials::ValidatorCredentials;
 use super::{address::Address, base64::Base64};
+use crate::error::Error;
 use async_graphql::*;
+use sui_indexer::apis::governance_api::exchange_rates;
 use sui_types::sui_system_state::sui_system_state_summary::SuiValidatorSummary as NativeSuiValidatorSummary;
-
 #[derive(Clone, Debug)]
 pub(crate) struct Validator {
     pub validator_summary: NativeSuiValidatorSummary,
     pub at_risk: Option<u64>,
     pub report_records: Option<Vec<Address>>,
     /// The checkpoint sequence number at which this was viewed at.
     pub checkpoint_viewed_at: u64,
+    /// The epoch at which this validator's information was requested to be viewed at.
+    pub requested_for_epoch: Option<u64>,
+}
+
+#[async_trait::async_trait]
+impl Loader<Option<u64>> for Db {
+    type Value = BTreeMap<NativeSuiAddress, (u64, Vec<(EpochId, PoolTokenExchangeRate)>)>;
+    type Error = Error;
+
+    async fn load(
+        &self,
+        keys: &[Option<u64>],
+    ) -> Result<
+        HashMap<Option<u64>, BTreeMap<NativeSuiAddress, (u64, Vec<(u64, PoolTokenExchangeRate)>)>>,
+        Error,
+    > {
+        let latest_sui_system_state = self
+            .inner
+            .spawn_blocking(move |this| this.get_latest_sui_system_state())
+            .await
+            .map_err(|_| Error::Internal("Failed to fetch latest Sui system state".to_string()))?;
+        let governance_api = GovernanceReadApi::new(self.inner.clone());
+        let exchange_rates = exchange_rates(&governance_api, &latest_sui_system_state)
+            .await
+            .map_err(|e| Error::Internal(format!("Error fetching exchange rates. {e}")))?;
+        let mut results = BTreeMap::new();
+
+        tracing::info!("Keys {:?}", keys);
+        // latest sui system state epoch is the current running one,
+        // so we need to subtract 1 to get the previous epoch
+        let requested_epoch = if let Some(epoch) = keys.first() {
+            epoch.unwrap_or_else(|| latest_sui_system_state.epoch - 1)
+        } else {
+            latest_sui_system_state.epoch - 1
+        };
```

Comment:
is this so for some exchange rate, we'll always have at least 2 entries? but what if the validator only has 1 exchange rate for the latest sui system state?

Reply:
Then you will get an `apy = 0.0`

Reply:
> is this so for some exchange rate, we'll always have at least 2 entries?

I believe we are subtracting 1 here because the exchange rate for the current epoch won't be available until the end of the current epoch, so the latest exchange rate we can get is actually from the epoch before the current one. Thus we need the `-1` here

Reply:
ah yeah that makes sense. now the comments make sense too ... but maybe we can update the comments a bit with that explanation?